src/sinks/util/concurrent_map.rs

```diff
@@ -0,0 +1,104 @@
+use std::{
+    future::Future,
+    num::NonZeroUsize,
+    panic,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures_util::{
+    ready,
+    stream::{Fuse, FuturesOrdered},
+    Stream, StreamExt,
+};
+use pin_project::pin_project;
+use tokio::task::JoinHandle;
+
+#[pin_project]
+pub struct ConcurrentMap<St, T>
```

Comment:
Possible candidate for `vector_core::stream`?

Reply:
Yeah, most certainly.

Reply:
Going to punt on moving any of this code until after the PR is merged, as it's grown a bit long in the tooth and I want to merge it sooner rather than later.